File: src/main/org/nlogo/prim/gui/_moviestatus.java
package org.nlogo.prim.gui ;

import org.nlogo.api.LogoException;
import org.nlogo.nvm.EngineException;
import org.nlogo.nvm.Syntax;
import org.nlogo.window.GUIWorkspace;

public final strictfp class _moviestatus
	extends org.nlogo.nvm.Reporter
{
	@Override
	public Syntax syntax()
	{
		return Syntax.reporterSyntax( Syntax.TYPE_STRING ) ;
	}
	@Override
	public Object report( final org.nlogo.nvm.Context context ) throws LogoException
	{
		if( ! ( workspace instanceof GUIWorkspace ) )
		{
			throw new EngineException(
				context , this , token().name() + " can only be used in the GUI" ) ;
		}
		org.nlogo.awt.MovieEncoder encoder = 
			( (GUIWorkspace) workspace ).movieEncoder;
			
		if ( encoder == null )
		{
			return "No movie.";
		}
		String status =  "" + encoder.getNumFrames() + " frames"
			+ "; frame rate = " + encoder.getFrameRate() ;
		
		if ( encoder.isSetup() ) 
		{
			java.awt.Dimension size = encoder.getFrameSize();
			status += "; size = " + size.width + "x" + size.height;
		}
		return status;
	}
}

File: src/main/org/nlogo/prim/gui/_moviestatus.scala
package org.nlogo.prim.gui

import org.nlogo.api.LogoException
import org.nlogo.awt.MovieEncoder
import org.nlogo.nvm.{ Context, EngineException, Reporter, Syntax }
import org.nlogo.window.GUIWorkspace

class _moviestatus extends Reporter {
  override def syntax =
    Syntax.reporterSyntax(Syntax.TYPE_STRING)
  override def report(context: Context) = {
    workspace match {
      case gw: GUIWorkspace =>
        status(gw.movieEncoder)
      case _ =>
        throw new EngineException(
          context, this, token.name + " can only be used in the GUI")
    }
  }
  private def status(encoder: MovieEncoder) =
    if(encoder == null)
      "No movie."
    else {
      val builder = new StringBuilder
      builder ++= encoder.getNumFrames + " frames" + "; "
      builder ++= "frame rate = " + encoder.getFrameRate
      if (encoder.isSetup) {
        val size = encoder.getFrameSize
        builder ++= "; size = " + size.width + "x" + size.height
      }
      builder.toString
    }
}
